### PR TITLE
added error handling

### PR DESCRIPTION
**Pull request by ajaypj**

### CODE REVIEW

File: guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java [hunk 1]
```diff
@@ -126,10 +139,28 @@ public void sendHeaders(final Metadata responseHeaders) {
...
+//                    if (!status.isOk()){
+//                        grpcFilters.forEach(filter -> filter.doHandleException(new StatusRuntimeException(status,trailers)));
+//                    }
+//                    super.close(status, trailers);
+//                } catch (Exception e) {
+//                    handleException(call, grpcFilters, e);
+//                } finally {
+//                    detachContext(contextWithHeaders, previous);    // detach headers from gRPC context
+//                }
+//            }
```
Comment (ajaypj) [inline guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java:163]: with close() we can get the exact Status and print it in the access logs.
Need to check how to get this with onCancel() because close() is not guaranteed to be called always in the javadocs.

File: guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java [hunk 2]
```diff
@@ -168,10 +200,10 @@ public void onMessage(Req request) {
             public void onCancel() {
                 Context previous = attachContext(contextWithHeaders);   // attaching headers to gRPC context
                 try {
+                    grpcFilters.forEach(filter -> filter.doHandleException(new Exception()));
```
Comment (ajaypj) [inline guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java:203]: any better way to get more specific exception/grpc status

File: guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java [hunk 3]
```diff
@@ -169,9 +199,23 @@ public void onCancel() {
                 Context previous = attachContext(contextWithHeaders);   // attaching headers to gRPC context
                 try {
                     super.onCancel();
-                } catch (RuntimeException ex) {
-                    handleException(call, ex);
-                    grpcFilters.forEach(filter -> filter.doHandleException(ex));
+                    grpcFilters.forEach(filter -> filter.doHandleException(new Exception()));
```
Comment (ajaypj) [inline guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java:202]: check how to set more descriptive exception containing grpc status code here, like DEADLINE_EXCEEDED

File: guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java [hunk 4]
```diff
@@ -169,9 +199,23 @@ public void onCancel() {
...
+            }
+
+            @Override
+            public void onComplete() {
+                Context previous = attachContext(contextWithHeaders);   // attaching headers to gRPC context
+                try {
+                    super.onComplete();
+                    grpcFilters.forEach(Filter::doEndFilter);
+                } catch (Exception e) {
+                    handleException(call, grpcFilters, e);
```
Comment (kingster) [inline guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java:218]: we shouldn't be calling handleException here. HandleException is for upstream. For Filter thrown error that needs to be returned without giving it back to the same filter.

File: guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java [hunk 5]
```diff
@@ -169,9 +199,23 @@ public void onCancel() {
                 Context previous = attachContext(contextWithHeaders);   // attaching headers to gRPC context
                 try {
                     super.onCancel();
-                } catch (RuntimeException ex) {
-                    handleException(call, ex);
-                    grpcFilters.forEach(filter -> filter.doHandleException(ex));
+                    grpcFilters.forEach(filter -> filter.doHandleException(new Exception()));
+                    grpcFilters.forEach(Filter::doEndFilter);
+                } catch (Exception e) {
+                    handleException(call, grpcFilters, e);
```
Comment (kingster) [inline guice/src/main/java/com/flipkart/gjex/grpc/interceptor/FilterInterceptor.java:205]: Same here, we shouldn't be calling handleException here. HandleException is for upstream. For Filter thrown error that needs to be returned without giving it back to the same filter.

File: core/src/main/java/com/flipkart/gjex/core/filter/Filter.java [hunk 6]
```diff
@@ -63,4 +63,9 @@ public void doProcessResponse(Res response) {}
      * @param e The exception that occurred.
      */
     public void doHandleException(Exception e) {}
+
+    /**
+     * Method for performing ending functions for a filter.
+     */
+    public void doEndFilter() {}
```
Comment (kingster) [inline core/src/main/java/com/flipkart/gjex/core/filter/Filter.java:70]: Not in favor of this. We already have an `destroy` method. Lifecycle methods should be used only for lifecycle and not for logic.

File: guice/src/main/java/com/flipkart/gjex/http/interceptor/HttpFilterInterceptor.java [hunk 7]
```diff
@@ -91,19 +92,23 @@ public final void doFilter(ServletRequest request, ServletResponse response,
                     .stream().collect(Collectors.toMap(String::toLowerCase, httpServletResponse::getHeader));
                 filters.forEach(filter -> filter.doProcessResponseHeaders(responseHeaders));
                 responseHeaders.forEach(responseWrapper::setHeader);
+                filters.forEach(filter -> filter.doProcessResponse(responseWrapper));
```
Comment (kingster) [inline guice/src/main/java/com/flipkart/gjex/http/interceptor/HttpFilterInterceptor.java:95]: This will cause filter thrown exception to be caught. We don't want that. Filter thrown error should not be caught

File: core/src/main/java/com/flipkart/gjex/core/filter/Filter.java (hunk 6 above)
Comment (ajaypj) [reply core/src/main/java/com/flipkart/gjex/core/filter/Filter.java:70]: This is not for ending lifecycle. This is for printing logs once only.

File: core/src/main/java/com/flipkart/gjex/core/filter/grpc/AccessLogGrpcFilter.java [hunk 8]
```diff
@@ -57,7 +58,7 @@ public class AccessLogGrpcFilter<R extends GeneratedMessageV3, S extends Generat
     private static final Logger logger = Logging.loggerWithName("ACCESS-LOG");
 
     public AccessLogGrpcFilter() {
-
+        accessLogContextBuilder.requestTime(System.currentTimeMillis());
```
Comment (kingster) [inline core/src/main/java/com/flipkart/gjex/core/filter/grpc/AccessLogGrpcFilter.java:61]: Init accessLogContextBuilder here and set `requestTime`

File: examples/build.gradle [hunk 9]
```diff
@@ -40,7 +40,10 @@ dependencies {
     implementation "io.dropwizard.metrics5:metrics-healthchecks:5.0.0"
     implementation "io.dropwizard.metrics5:metrics-annotation:5.0.0"
     implementation "com.palominolabs.metrics:metrics-guice:5.0.1"
-    implementation 'ru.vyarus:guice-validator:1.2.0'
+    implementation ('ru.vyarus:guice-validator:1.2.0') {
+        exclude group: 'com.google.inject', module: 'guice'
+    }
+    implementation 'com.google.inject:guice:5.1.0'
```
Comment (kingster) [inline examples/build.gradle:46]: Why are these required?

**Review by kingster — APPROVED** [review]

File: core/src/main/java/com/flipkart/gjex/core/filter/grpc/AccessLogGrpcFilter.java (hunk 8 above)
Comment (ajaypj) [reply core/src/main/java/com/flipkart/gjex/core/filter/grpc/AccessLogGrpcFilter.java:62]: done
